### PR TITLE
[NO MERGE] Refactoring and fix some warnings

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs [hunk 1]
```diff
@@ -259,7 +242,7 @@ private static List<ArgumentSyntax> GetReducedArguments(IMethodSymbol methodSymb
         }
 
         private static bool CheckForCastSimplification(IMethodSymbol outerMethodSymbol, InvocationExpressionSyntax outerInvocation,
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs]: MINOR Change return type to 'void'; not a single caller uses the returned value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3241)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs [hunk 2]
```diff
@@ -237,13 +237,15 @@ public static Accessibility GetEffectiveAccessibility(this ISymbol symbol)
...
-                    case Accessibility.Internal:
-                        result = Accessibility.Internal;
-                        continue;
+                    return Accessibility.Private;
+                }
+
+                if (container.DeclaredAccessibility == Accessibility.Internal)
+                {
+                    result = Accessibility.Internal;
+                    continue;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs]: MINOR Remove this redundant jump. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3626)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CatchRethrow.cs [hunk 3]
```diff
@@ -116,7 +116,7 @@ private void RaiseOnInvalidCatch(SyntaxNodeAnalysisContext context)
             return catches
                 .Select(clause =>
                     clause.Declaration?.Type != null
-                        ? semanticModel.GetTypeInfo(clause.Declaration?.Type).Type as INamedTypeSymbol
+                        ? semanticModel.GetTypeInfo(clause.Declaration.Type).Type as INamedTypeSymbol
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CatchRethrow.cs]: Why?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DisposeNotImplementingDispose.cs [hunk 4]
```diff
@@ -51,7 +51,7 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                 {
                     var declaredSymbol = (INamedTypeSymbol)c.Symbol;
 
-                    if (declaredSymbol.DeclaringSyntaxReferences.Count() > 1)
+                    if (declaredSymbol.DeclaringSyntaxReferences.Length > 1)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DisposeNotImplementingDispose.cs]: This should go in the other commit.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassName.cs [hunk 5]
```diff
@@ -68,7 +68,7 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                     }
 
                     string suggestion;
-                    if (TryGetChangedName(identifier.ValueText, typeDeclaration, c.SemanticModel.Compilation.IsTest(), out suggestion))
+                    if (TryGetChangedName(identifier.ValueText, typeDeclaration, c.IsTest(), out suggestion))
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassName.cs]: You should delete the old `IsTest` on Compilation if it is not used anymore. (disregard if you already did it)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/FieldShouldBeReadonly.cs [hunk 6]
```diff
@@ -77,7 +77,7 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                         return;
                     }
 
-                    if (declaredSymbol.DeclaringSyntaxReferences.Count() > 1)
+                    if (declaredSymbol.DeclaringSyntaxReferences.Length > 1)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/FieldShouldBeReadonly.cs]: This should go into the other commit.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 7]
```diff
@@ -132,19 +132,22 @@ private static bool IsInCheckedContext(ExpressionSyntax expression, SemanticMode
                     ? model.GetSymbolInfo(constructorOrMethod).Symbol as IMethodSymbol
                     : null;
 
-                return (methodSymbol != null && argumentIndex.Value < methodSymbol.Parameters.Length)
-                    ? methodSymbol.Parameters[argumentIndex.Value].Name.SplitCamelCaseToWords()
-                        .Any(name => checkedVariableNames.Contains(name))
-                    : false;
+                return methodSymbol != null &&
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: Good improvement!

I'd move these two changes to a separate commit as well.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs [hunk 8]
```diff
@@ -467,7 +433,7 @@ private static bool MethodIsNotUsingPredicate(IMethodSymbol methodSymbol, Invoca
         {
             var arguments = GetReducedArguments(methodSymbol, invocation);
 
-            return !arguments.Any() && MethodNamesWithPredicate.Contains(methodSymbol.Name);
+            return arguments.Count == 0 && MethodNamesWithPredicate.Contains(methodSymbol.Name);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs]: You are really obsessed with performance, eh? :) I am curious if it will be much faster in real world if we first check arguments.Count and then call GetReducedArguments...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs [hunk 9]
```diff
@@ -194,20 +194,15 @@ private static SyntaxNode GetUnreachableStatement(IEnumerable<ExpressionSyntax>
...
+            return constantValue
+                ? (SyntaxNode)
                     (parent as ConditionalExpressionSyntax)?.WhenFalse ??
-                    (parent as IfStatementSyntax)?.Else?.Statement;
-            }
-            else
-            {
-                return (SyntaxNode)
+                    (parent as IfStatementSyntax)?.Else?.Statement
+                : (SyntaxNode)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: I don't like nested ternary operators

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/TooManyInternalStatesException.cs [hunk 10]
```diff
@@ -19,10 +19,26 @@
  */
 
 using System;
+using System.Runtime.Serialization;
 
 namespace SonarAnalyzer.Helpers.FlowAnalysis.Common
 {
+    [Serializable]
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/TooManyInternalStatesException.cs]: I doubt this exception will be able to bubble up outside of the method where it is caught :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CatchRethrow.cs (hunk 3 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CatchRethrow.cs]: Because we test `clause.Declaration?.Type != null` before so when we reach this branch we know that `clause.Declaration.Type` isn't null (and so `clause.DeclarationType` isn't).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 7 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: Ok will do.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassName.cs (hunk 5 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassName.cs]: I didn't.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs (hunk 9 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs]: I kept it because there was quite a lot in common but happy to undo if you prefer.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/TooManyInternalStatesException.cs (hunk 10 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/TooManyInternalStatesException.cs]: You are right the scope is small now but I'd rather fix the code than mark a FP on this case on Next.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionQuerySimplification.cs]: MINOR Change return type to 'void'; not a single caller uses the returned value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3241)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs (hunk 2 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs]: MINOR Remove this redundant jump. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3626)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs (hunk 2 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs]: MINOR Remove this redundant jump. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3626)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [MethodsShouldUseBaseTypes.cs#L129](https://github.com/SonarSource/sonar-csharp/blob/7555d0864bbdf6653d7df507f8128f07d8a84689/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs#L129): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
